crates/sui-graphql-rpc/src/context_data/pg_backend.rs

```diff
@@ -362,6 +364,105 @@ impl GenericQueryBuilder<Pg> for PgQueryBuilder {
 
         query
     }
+    fn multi_get_events(
+        cursor: Option<(i64, i64)>,
+        descending_order: bool,
+        limit: i64,
+        filter: Option<EventFilter>,
+    ) -> Result<events::BoxedQuery<'static, Pg>, Error> {
+        let mut query = events::dsl::events.into_boxed();
+        if let Some(cursor) = cursor {
+            if descending_order {
+                query = query.filter(
+                    events::dsl::tx_sequence_number.lt(cursor.0).or(
+                        events::dsl::tx_sequence_number
+                            .eq(cursor.0)
+                            .and(events::dsl::event_sequence_number.lt(cursor.1)),
+                    ),
+                );
+            } else {
+                query = query.filter(
+                    events::dsl::tx_sequence_number.gt(cursor.0).or(
+                        events::dsl::tx_sequence_number
+                            .eq(cursor.0)
+                            .and(events::dsl::event_sequence_number.gt(cursor.1)),
+                    ),
+                );
+            }
+        }
+        if descending_order {
+            query = query
+                .order(events::dsl::tx_sequence_number.desc())
+                .then_order_by(events::dsl::event_sequence_number.desc());
+        } else {
+            query = query
+                .order(events::dsl::tx_sequence_number.asc())
+                .then_order_by(events::dsl::event_sequence_number.asc());
+        }
+
+        query = query.limit(limit + 1);
+
+        if let Some(filter) = filter {
+            if let Some(sender) = filter.sender {
+                // Construct a subquery to filter on senders - this is because we do not have an index on the senders column.
+                let subquery = tx_senders::dsl::tx_senders
+                    .filter(tx_senders::dsl::sender.eq(sender.into_vec()))
+                    .select(tx_senders::dsl::tx_sequence_number);
+
+                query = query.filter(events::dsl::tx_sequence_number.eq_any(subquery));
+            }
```

Comment:
You may have looked into this already but is there any appreciable difference in query performance between writing the query like this, vs using a join?

e.g. this is currently:

```sql
SELECT *
FROM   event
WHERE  tx_sequence_number IN (
    SELECT tx_sequence_number 
    FROM   tx_senders 
    WHERE  sender = $SENDER
)
```

As opposed to something like:

```sql
SELECT events.*
FROM   events INNER JOIN tx_senders USING (tx_sequence_number)
WHERE  sender = $SENDER
GROUP BY (tx_sequence_number, event_sequence_number)
```

Reply:
I don't think the group by is necessary for the latter query. The query planner comes up with a very similar execution plan and cost for the two queries above. I did start with trying to write the query as an inner join through dsl query builder, but ran into some issues (that I have since forgotten) and ended up with the subquery syntax instead. Carried it over here.

Reply:
No worries, was just curious!  On the group by, I think it's not necessary today, but will be when we start supporting multiple senders for a transaction (which is a while away, but still).